test/e2e/self_hosted.go

```diff
@@ -215,6 +259,114 @@ func SelfHostedSpec(ctx context.Context, inputGetter func() SelfHostedSpecInput)
 			return matchUnstructuredLists(preMoveMachineList, postMoveMachineList)
 		}, "3m", "30s").Should(BeTrue(), "Machines should not roll out after move to self-hosted cluster")
 
+		if !input.EnableUpgrade {
+			// Only do upgrade step if defined by test input.
+			return
+		}
+
+		By("Upgrading the self-hosted Cluster")
+
+		if clusterResources.Cluster.Spec.Topology != nil {
+			clusterModifiers := []func(c *clusterv1.Cluster){}
+
+			if isDockerCluster {
+				images := []string{}
+				for _, image := range input.E2EConfig.Images {
+					images = append(images, fmt.Sprintf("%q", image.Name))
+				}
+				preLoadVariable := clusterv1.ClusterVariable{
+					Name: "preLoadImages",
+					Value: apiextensionsv1.JSON{
+						Raw: []byte(`[` + strings.Join(images, ",") + `]`),
+					},
+				}
+
+				clusterModifiers = append(clusterModifiers, func(c *clusterv1.Cluster) {
+					c.Spec.Topology.Variables = append(c.Spec.Topology.Variables, preLoadVariable)
+				})
```

Comment:
This solution currently leads to having a parallel starting rollout for KCP and MD:

* KCP: version + preLoadImages change on DockerMachineTemplate
* MD: preLoadImages change on DockerMachineTemplate only

After KCP finished, the MD gets updated to also have the version change.

On long-term I think it would be best to be able to set the variable for preLoadImages already on Cluster Creation. 
- xref: https://github.com/kubernetes-sigs/cluster-api/pull/7244/files#diff-7084fa27f2b10b8d2472295c14187d21ea186aae39efa12e051f340562a4cca4R128

Reply:
adjusted implementation to be able to pass variables for cluster creation.

Reply:
Looks good!